modules/transport-netty4/src/main/java/org/elasticsearch/transport/CopyBytesSocketChannel.java

```diff
@@ -162,9 +162,17 @@ private void adjustMaxBytesPerGatheringWrite(int attempted, int written, int old
     private static void copyBytes(ByteBuffer[] source, int nioBufferCnt, ByteBuffer destination) {
         for (int i = 0; i < nioBufferCnt && destination.hasRemaining(); i++) {
             ByteBuffer buffer = source[i];
-            assert buffer.hasArray() : "Buffer must have heap array";
             int nBytesToCopy = Math.min(destination.remaining(), buffer.remaining());
-            destination.put(buffer.array(), buffer.arrayOffset() + buffer.position(), nBytesToCopy);
+            if (buffer.hasArray()) {
+                destination.put(buffer.array(), buffer.arrayOffset() + buffer.position(), nBytesToCopy);
+            } else {
+                int initialLimit = buffer.limit();
```

Comment:
Netty uses static direct byte buffers to chunked transfer encoded line breaks. So we have to handle them.